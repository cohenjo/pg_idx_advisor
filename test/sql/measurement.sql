-- test for target list and inherited tables

CREATE OR REPLACE FUNCTION measurement_insert_trigger()
RETURNS TRIGGER AS $$
BEGIN
    IF ( NEW.logdate >= DATE '2006-02-01' AND
         NEW.logdate < DATE '2006-03-01' ) THEN
        INSERT INTO measurement_y2006m02 VALUES (NEW.*);
    ELSIF ( NEW.logdate >= DATE '2006-03-01' AND
            NEW.logdate < DATE '2006-04-01' ) THEN
        INSERT INTO measurement_y2006m03 VALUES (NEW.*);
    ELSE
        RAISE EXCEPTION 'Date out of range.  Fix the measurement_insert_trigger() function!';
    END IF;
    RETURN NULL;
END;
$$ language plpgsql;

CREATE TABLE measurement (
    city_id         int not null,
    logdate         date not null,
    peaktemp        int,
    unitsales       int
);

CREATE TABLE measurement_y2006m02 (
    CHECK ( logdate >= DATE '2006-02-01' AND logdate < DATE '2006-03-01' )
) INHERITS (measurement);

CREATE TABLE measurement_y2006m03 (
    CHECK ( logdate >= DATE '2006-03-01' AND logdate < DATE '2006-04-01' )
) INHERITS (measurement);

CREATE TRIGGER insert_measurement_trigger
    BEFORE INSERT ON measurement
    FOR EACH ROW EXECUTE PROCEDURE measurement_insert_trigger();

create index idx_measurement_unitsales_y2006m02 ON measurement_y2006m02 (unitsales);

-- now load the advisor
load 'pg_idx_advisor.so';
\o /tmp/pg_idx_tst.out
-- we should get advices on both parent and cheild tables
explain select max(unitsales) from measurement;
\o
select attrs,benefit,indclass,indoption,query,recommendation from index_advisory;

