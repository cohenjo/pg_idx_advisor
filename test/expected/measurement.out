-- test for target list and inherited tables
CREATE OR REPLACE FUNCTION measurement_insert_trigger()
RETURNS TRIGGER AS $$
BEGIN
    IF ( NEW.logdate >= DATE '2006-02-01' AND
         NEW.logdate < DATE '2006-03-01' ) THEN
        INSERT INTO measurement_y2006m02 VALUES (NEW.*);
    ELSIF ( NEW.logdate >= DATE '2006-03-01' AND
            NEW.logdate < DATE '2006-04-01' ) THEN
        INSERT INTO measurement_y2006m03 VALUES (NEW.*);
    ELSE
        RAISE EXCEPTION 'Date out of range.  Fix the measurement_insert_trigger() function!';
    END IF;
    RETURN NULL;
END;
$$ language plpgsql;
CREATE TABLE measurement (
    city_id         int not null,
    logdate         date not null,
    peaktemp        int,
    unitsales       int
);
CREATE TABLE measurement_y2006m02 (
    CHECK ( logdate >= DATE '2006-02-01' AND logdate < DATE '2006-03-01' )
) INHERITS (measurement);
CREATE TABLE measurement_y2006m03 (
    CHECK ( logdate >= DATE '2006-03-01' AND logdate < DATE '2006-04-01' )
) INHERITS (measurement);
CREATE TRIGGER insert_measurement_trigger
    BEFORE INSERT ON measurement
    FOR EACH ROW EXECUTE PROCEDURE measurement_insert_trigger();
create index idx_measurement_unitsales_y2006m02 ON measurement_y2006m02 (unitsales);
-- now load the advisor
load 'pg_idx_advisor.so';
NOTICE:  IND ADV: plugin loaded
\o /tmp/pg_idx_tst.out
-- we should get advices on both parent and cheild tables
explain select max(unitsales) from measurement;
INFO:  
** Plan with Original indexes **

\o
select attrs,benefit,indclass,indoption,query,recommendation from index_advisory;
 attrs | benefit  | indclass | indoption |                   query                    |                 recommendation                  
-------+----------+----------+-----------+--------------------------------------------+-------------------------------------------------
 {1}   |  21.9635 | {1978}   | {1978}    | select * from t where a = 100;             | create index on t(a)
 {2}   |  21.9635 | {1978}   | {1978}    | select * from t where b = 100;             | create index on t(b)
 {1}   |  14.8048 | {1978}   | {1978}    | select * from t where a = 100 and b = 100; | create index on t(a)
 {2}   |  14.8048 | {1978}   | {1978}    | select * from t where a = 100 and b = 100; | create index on t(b)
 {1}   |  11.7747 | {1978}   | {1978}    | select * from t where a = 100 or b = 100;  | create index on t(a)
 {2}   |  11.7747 | {1978}   | {1978}    | select * from t where a = 100 or b = 100;  | create index on t(b)
 {4}   | 0.066363 | {1978}   | {1978}    | select max(unitsales) from measurement;    | create index on measurement(unitsales)
 {4}   | 0.066363 | {1978}   | {1978}    | select max(unitsales) from measurement;    | create index on measurement_y2006m03(unitsales)
(8 rows)

